modules/code-deployment/src/main/java/org/apache/ignite/internal/deployunit/DeploymentManagerImpl.java

```diff
@@ -0,0 +1,424 @@
+     * @param configuration Deployment configuration.
+     * @param cmgManager Cluster management group manager.
+     */
+    public DeploymentManagerImpl(ClusterService clusterService,
+            MetaStorageManager metaStorage,
+            Path workDir,
+            DeploymentConfiguration configuration,
+            ClusterManagementGroupManager cmgManager) {
+        this.clusterService = clusterService;
+        this.metaStorage = metaStorage;
+        this.configuration = configuration;
+        this.cmgManager = cmgManager;
+        unitsFolder = workDir;
+    }
+
+    @Override
+    public CompletableFuture<Boolean> deployAsync(String id, Version version, DeploymentUnit deploymentUnit) {
+        if (id == null || id.isBlank() || version == null) {
+            return CompletableFuture.failedFuture(new DeploymentUnitIdentifierException());
+        }
+
+        ByteArray key = new ByteArray(UNITS_PREFIX + id + ":" + version.render());
+
+        UnitMeta meta = new UnitMeta(id, version, deploymentUnit.name(), Collections.emptyList());
+
+        Operation put = put(key, UnitMetaSerializer.serialize(meta));
+
+        DeployUnitRequestBuilder builder = DeployUnitRequestImpl.builder();
+
+        try {
+            builder.unitContent(deploymentUnit.content().readAllBytes());
+        } catch (IOException e) {
+            LOG.error("Error to read deployment unit content", e);
+            return CompletableFuture.failedFuture(new DeploymentUnitReadException(e));
+        }
+        DeployUnitRequest request = builder
+                .unitName(deploymentUnit.name())
+                .id(id)
+                .version(version.render())
+                .build();
+
+        return metaStorage.invoke(notExists(key), put, Operations.noop())
+                .thenCompose(success -> {
+                    if (success) {
+                        return doDeploy(request);
+                    }
+                    LOG.error("Failed to deploy meta of unit " + id + ":" + version);
+                    return CompletableFuture.failedFuture(new DeployUnitWriteMetaException());
+                })
+                .thenApply(completed -> {
+                    if (completed) {
+                        startDeployAsyncToCmg(request);
+                    }
+                    return completed;
+                });
+    }
+
+    private void startDeployAsyncToCmg(DeployUnitRequest request) {
+        cmgManager.cmgNodes()
+                .thenApply(nodes -> nodes.stream().map(node -> clusterService.topologyService().getByConsistentId(node)).collect(
+                        Collectors.toList()))
+                .thenAccept(clusterNodes -> {
+                    for (ClusterNode clusterNode : clusterNodes) {
+                        inFlightFutures.registerFuture(
+                                clusterService.messagingService()
+                                        .invoke(clusterNode, request, Long.MAX_VALUE)
```

Comment:
This code looks too nested. Maybe it would make sense to extract some methods to make it more readable.